### DISCUSSION TITLE
Asynchronous command invocation interface not send all commands when connection closed

### DISCUSSION THREAD

**Discussion by Sinbol** [redis/lettuce#2951] [Q&A]
I tried to write a simple demo using the async way to call the set command and found that when the program ended, Redis did not actually receive the command. I am curious why the connection closure does not guarantee the complete execution of the command.

i dont want to handle the response, so the simple code like bellow

```java
public class RedisAsync {
    public static void main(String[] args1) throws ExecutionException, InterruptedException {
        ArrayList<RedisURI> list = new ArrayList <RedisURI>();
        list.add(RedisURI.create({your_addr_info}));
        RedisClusterClient redisClient = RedisClusterClient.create(list);

        StatefulRedisClusterConnection<String, String> connection = redisClient.connect();
        RedisCodec<String, String> codec = StringCodec.UTF8;
        RedisAdvancedClusterAsyncCommands<String, String> async = connection.async();

        long val = 0;
        while (val++ < 100) {
            UUID uuid = UUID.randomUUID();
            long hash = uuid.getMostSignificantBits();
            String key_name = String.valueOf(hash);

            try {
                async.set(key_name, String.valueOf(val+1000));
            } catch (Exception e) {
                System.out.println("error occur");
            }
        }
        redisClient.shutdown();
    }
}
```

Additionally, maybe helpful that if sleep() for some time before return, it can help sometime.
I guess that the client closed too quickly before sending the command, but is there any interface to ensure  a safe shutdown?  i also tried the async.flushCommands() before shutdown, but not help... kindly help

**Comment by tishun** [comment]
Hey @Sinbol ,

Have you tried to call `.get()` on the future returned by `async.set()` ?

```java
               async.set(key_name, String.valueOf(val+1000)).get();
```

**Reply by Sinbol** [reply]
@tishun I have tried, but give up for efficiency reasons, I prefer not to do this because I don't care about the return value. However, thank you very much for your answer， as you told， now I plan to wait for the last command to return before the connection is closed.

**Reply by tishun** [reply]
> @tishun I have tried, but give up for efficiency reasons, I prefer not to do this because I don't care about the return value. However, thank you very much for your answer， as you told， now I plan to wait for the last command to return before the connection is closed.

Apologies, not sure what you mean by "efficiency reasons", could you clarify?

**Reply by Sinbol** [reply]
I meant calling .get() caused ops going lower during my test， I tried to have a better ops performance.

**Reply by tishun** [reply]
Using your example I can't reproduce the same issue, I can fetch all the results from the server after the application finishes.

How do you verify that the command is not received?
